frame/balances/src/lib.rs

```diff
@@ -1149,15 +1149,19 @@ impl<T: Config<I>, I: 'static> fungible::Transfer<T::AccountId> for Pallet<T, I>
 
 impl<T: Config<I>, I: 'static> fungible::Unbalanced<T::AccountId> for Pallet<T, I> {
 	fn set_balance(who: &T::AccountId, amount: Self::Balance) -> DispatchResult {
-		Self::mutate_account(who, |account| {
-			account.free = amount;
+		Self::mutate_account(who, |account| -> DispatchResult {
+			// fungibles::Unbalanced::decrease_balance didn't check account.reserved
+			// free = new_balance - reserved
+			account.free =
+				amount.checked_sub(&account.reserved).ok_or(ArithmeticError::Underflow)?;
```

Comment:
This assumes that `Unbalanced::set_balance` is always being used with an amount derived from `Inspect::balance()`. This is incorrect.
It would be more correct to separately get/set both `free` and `reserve`. within their respective traits. Then `InspectMutate` could have a function for getting `free + reserve` when it's needed for `minimum_balance`/ED requirements

Reply:
The trait doesn't really concern itself with "free" and "reserved" - that's a distinction only relevant for the pallet and the `*Hold` traits. I would avoid polluting the base traits with the concern of hold/reserve, as not all that many uses care about them.

In short, I think this is a perfectly reasonable fix.

Reply:
I'd argue that because of the `*Hold` traits that any actions that account for `reserve` should only belong to some `*Hold` trait and that other fungible(s) traits should operate correctly without being concerned with `reserve`. 
Even if this means having `*Hold` traits that overlap some of the functionalities from trait functions that aren't dependent on `*Hold`.

More plainly, `*Hold` functionality should be an extension of fungible(s) traits and not something at all implemented whiten the regular fungible(s) traits such as `Mutate`

Reply:
If I'm not mistaken, the purpose of this function is to set the total balance of the supplied account, irrespective of whether the end result is made up of free balances, reserve balances or a mixture of both. The fact that an account previously had held funds should not automatically mean that they suddenly needed to call another function just to modify the total balance -- that would mean users need to keep track of the state of the account in order to call the correct function.

I get the part where it is functionally cleaner if only `*Hold` traits deal with reserve balances, but it is highly inconvenient and an unnecessary distinction from a usability perspective.